### PR TITLE
hr-adaptivity with TMOP-based error estimators

### PR DESCRIPTION
**Pull request by kmittal2**
This PR enables hr-adaptivity using TMOP-based error estimators. Below is an example that shows a 4x4 mesh (left) optimized using hr-adaptivity with (center) isotropic refinement for size adaptivity, and (right) isotropic and anisotropic refinements for size and aspect-ratio adaptivity.

4x4 isotropic anisotropic

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1922](https://github.com/mfem/mfem/pull/1922) | @kmittal2 | @tzanio | @vladotomov + @tzanio + @rw-anderson | 12/15/20 | 08/01/21 | 08/08/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Can you merge `master` here and resolve the conflicts please?

File: mesh/ncmesh.hpp [hunk 1]
```diff
@@ -145,7 +148,10 @@ class NCMesh
        Note that if anisotropic refinements are present in the mesh, some of the
        derefinements may have to be skipped to preserve mesh consistency. */
    virtual void Derefine(const Array<int> &derefs);
+   virtual void GetFineToCoarsePartitioning(const Array<int> &derefs,
+                                            Array<int> &new_ranks) {}
 
+   int GetNGhostElements() const { return 0; }
```
Comment (vladotomov) [inline mesh/ncmesh.hpp:154]: Is this needed in the serial class? Seems that you only use the parallel class version.

File: mesh/ncmesh.hpp [hunk 2]
```diff
@@ -145,7 +148,10 @@ class NCMesh
        Note that if anisotropic refinements are present in the mesh, some of the
        derefinements may have to be skipped to preserve mesh consistency. */
    virtual void Derefine(const Array<int> &derefs);
+   virtual void GetFineToCoarsePartitioning(const Array<int> &derefs,
+                                            Array<int> &new_ranks) {}
```
Comment (vladotomov) [inline mesh/ncmesh.hpp:152]: This seems to be relevant only in parallel. It could be defined only in the parallel class.

File: mesh/ncmesh.hpp [hunk 3]
```diff
@@ -70,6 +70,9 @@ struct CoarseFineTransformations
                            Table &ref_type_to_matrix,
                            Array<Geometry::Type> &ref_type_to_geom) const;
 
+   void GetCoarseToFineMap(const Mesh &fine_mesh,
+                           Table &coarse_to_fine, int NGhost=0) const;
```
Comment (vladotomov) [inline mesh/ncmesh.hpp:74]: Could you add a comment about the difference with the existing version?

File: mesh/pncmesh.hpp [hunk 4]
```diff
@@ -86,6 +86,11 @@ class ParNCMesh : public NCMesh
        in sync. The interface is identical. */
    virtual void Derefine(const Array<int> &derefs);
 
+   /** Gets partitioning for the coarse mesh if the current fine mesh were to
+       be derefined. */
+   virtual void GetFineToCoarsePartitioning(const Array<int> &derefs,
+                                            Array<int> &new_ranks);
```
Comment (vladotomov) [inline mesh/pncmesh.hpp:92]: This function can be `const`.

File: mesh/pncmesh.cpp [hunk 5]
```diff
@@ -1483,6 +1483,43 @@ void ParNCMesh::LimitNCLevel(int max_nc_level)
...
+         int fine_rank = elements[leaf_elements[fine[j]]].rank;
+         coarse_rank = std::min(coarse_rank, fine_rank);
+      }
+      for (int j = 0; j < size; j++)
+      {
+         new_ranks[fine[j]] = coarse_rank;
+      }
+   }
+
+   int target_elements = 0;
```
Comment (vladotomov) [inline mesh/pncmesh.cpp:1516]: Is this `target_elements` some leftover debug computation?

File: miniapps/meshing/mesh-optimizer.hpp [hunk 6]
```diff
@@ -279,3 +386,959 @@ void DiffuseField(ParGridFunction &field, int smooth_steps)
...
+   if (pncmesh)
+   {
+      const Table &dreftable = pncmesh->GetDerefinementTable();
+      Array<int> drefs, new_ranks;
+      for (int i = 0; i < dreftable.Size(); i++)
+      {
+         drefs.Append(i);
+      }
+      pncmesh->GetFineToCoarsePartitioning(drefs, new_ranks);
+      new_ranks.SetSize(pmesh->GetNE());
```
Comment (vladotomov) [inline miniapps/meshing/mesh-optimizer.hpp:454]: Doesn't `new_ranks` come out of the `GetFineToCoarsePartitioning` call (above) with the correct size?

File: miniapps/meshing/mesh-optimizer.hpp [hunk 7]
```diff
@@ -279,3 +386,959 @@ void DiffuseField(ParGridFunction &field, int smooth_steps)
    delete Lap;
 }
 #endif
+
+
+/// This class is used to update GridFunction and FESpaces after mesh updates.
+/// It also provides functionality to reset essential boundary conditions, and
+/// rebalance ParMesh.
+class TMOPAMR
```
Comment (vladotomov) [inline miniapps/meshing/mesh-optimizer.hpp:394]: All these classes should be moved to `tmop_amr.hpp/cpp` in `/fem` s.t. they are available with the library.

**Review by vladotomov — CHANGES_REQUESTED** [review]

**Comment by tzanio** [comment]
Extending the review deadline to the end of the week...

**Review by jandrej — COMMENTED** [review]
I finished a first pass through the code and overall I can't find major roadblocks. I'll need some time to carefully go through all the new functionality though.

File: miniapps/meshing/mesh-optimizer.cpp [hunk 8]
```diff
@@ -221,6 +229,13 @@ int main(int argc, char *argv[])
                   "0: Use single metric\n\t"
                   "1: Shape + space-dependent size given analytically\n\t"
                   "2: Shape + adapted size given discretely; shared target");
+   args.AddOption(&hradaptivity, "-hr", "--hr-adaptivity", "-no-hr",
+                  "--no-hr-adaptivity",
+                  "Enable hr-adaptivity.");
+   args.AddOption(&amr_metric_id, "-hmid", "--h-metric",
+                  "same options as metric_id");
+   args.AddOption(&hessiantype, "-ht", "--Hessian type for hr-adaptivity examples",
+                  "1-3");
```
Comment (vladotomov) [inline miniapps/meshing/mesh-optimizer.cpp:238]: Expain the options (don't go into details, just something general that corresponds to the paper tests).

File: miniapps/meshing/mesh-optimizer.cpp [hunk 9]
```diff
@@ -221,6 +229,13 @@ int main(int argc, char *argv[])
                   "0: Use single metric\n\t"
                   "1: Shape + space-dependent size given analytically\n\t"
                   "2: Shape + adapted size given discretely; shared target");
+   args.AddOption(&hradaptivity, "-hr", "--hr-adaptivity", "-no-hr",
+                  "--no-hr-adaptivity",
+                  "Enable hr-adaptivity.");
+   args.AddOption(&amr_metric_id, "-hmid", "--h-metric",
+                  "same options as metric_id");
```
Comment (vladotomov) [inline miniapps/meshing/mesh-optimizer.cpp:236]: Comment on the usage of `hmid`

File: miniapps/meshing/mesh-optimizer.hpp [hunk 10]
```diff
@@ -117,79 +115,185 @@ void discrete_aspr_3d(const Vector &x, Vector &v)
    v[2] = l3/pow(l2*l1,0.5);
 }
 
+
 class HessianCoefficient : public TMOPMatrixCoefficient
 {
 private:
+   int dim;
    int metric;
+   int amr_type;
```
Comment (vladotomov) [inline miniapps/meshing/mesh-optimizer.hpp:124]: The name `amr` is confusing -- this has nothing to do with AMR, does it?

**Review by vladotomov — CHANGES_REQUESTED** [review]

File: fem/tmop.hpp [hunk 11]
```diff
@@ -1486,6 +1568,13 @@ class TMOP_Integrator : public NonlinearFormIntegrator
                                     ElementTransformation &T,
                                     const Vector &elfun, DenseMatrix &elmat);
 
+   TMOP_QualityMetric &GetAMRQualityMetric() { return *h_metric; }
+
+   void Update();
```
Comment (vladotomov) [inline fem/tmop.hpp:1573]: Should we name this `UpdateAfterMeshTopologyChange` to correspond to the other update function, and the function in `DiscreteTC`?

File: fem/tmop_amr.hpp [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+   void SetSpatialIndicatorCritical(double val_) { spat_gf_critical = val_; }
+
+   /// Reset the error estimator.
+   virtual void Reset() { current_sequence = -1; }
+};
+
+
+
+// TMOPRefiner is ThresholdRefiner with total_error_fraction = 0.;
+class TMOPRefiner : public ThresholdRefiner
```
Comment (vladotomov) [inline fem/tmop_amr.hpp:135]: Do we need a whole new class? Can't we just create a `ThresholdRefiner` and call `SetTotalErrorFraction(0.)` right away?

File: fem/tmop_amr.hpp [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+   void AddFESpaceForUpdate(ParFiniteElementSpace *pfes_)
+   {
+      pfespacearr.Append(pfes_);
+   }
+#endif
+
+#ifdef MFEM_USE_MPI
+   // Rebalance ParMesh such that all the children elements are moved to the same
+   // MPI rank where the parent will be if the mesh were to be derefined.
+   void RebalanceParNCMesh();
```
Comment (vladotomov) [inline fem/tmop_amr.hpp:279]: Does it need to be public?

File: fem/tmop_amr.cpp [hunk 14]
```diff
@@ -0,0 +1,947 @@
...
+                         ", Elements: " << NEGlob << std::endl;
+            }
+
+            if (!tmop_dr->Derefined() && tmop_r->Stop())
+            {
+               radaptivity = false;
+               if (myid == 0)
+               {
+                  mfem::out << "AMR stopping criterion satisfied. Stop.\n";
+                  break;
```
Comment (vladotomov) [inline fem/tmop_amr.cpp:721]: The `break` is inside the `if`, looks like an error.

File: fem/tmop_amr.cpp [hunk 15]
```diff
@@ -0,0 +1,947 @@
...
+               break;
+            }
+         } //n_h
+      } //n_hr
+   }
+   else
+   {
+#ifdef MFEM_USE_MPI
+      int NEGlob = pmesh->GetGlobalNE();
+      double tmopenergy = pnlf->GetParGridFunctionEnergy(*x);
```
Comment (vladotomov) [inline fem/tmop_amr.cpp:665]: Seems that this calculation is not needed - it gets recomputed below every time.

File: fem/tmop.hpp [hunk 16]
```diff
@@ -1184,6 +1215,33 @@ class DiscreteAdaptTC : public TargetConstructor
                                               const Vector &elfun,
                                               IsoparametricTransformation &Tpr,
                                               DenseTensor &dJtr) const;
+
+   // Generates tspec_vals for target construction using intrule
+   // Used for the refinement component in hr-adaptivity.
+   void SetTspecFromIntRule(int e_id, const IntegrationRule &intrule);
+
+   void SetMinSizeForTargets(double min_size_) { lim_min_size = min_size_; }
```
Comment (vladotomov) [inline fem/tmop.hpp:1223]: Explain the use of this.

**Review by vladotomov — APPROVED** [review]
Impressive work @kmittal2. This is a substantial amount of work that requires good understanding of the internal AMR specifics. Thank you!

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
@rw-anderson and @jandrej -- please chime in if you have additional comments.

**Comment by v-dobrev** [comment]
This sample run:
```console
mesh-optimizer -m stretched2D.mesh -o 2 -mid 2 -tid 1 -ni 50 -qo 5 -nor -vl 1 -alc 0.5 -no-vis
```
gives a different result in this PR:
```diff
-  Final strain energy: 0.00919635 = metrics: 0.00858393 + limiting term: 0.000612423
+  Final strain energy: 0.00919635 = metrics: 0.00919635 + limiting term: 0
```

**Comment by v-dobrev** [comment]
Also
```console
mesh-optimizer -m stretched2D.mesh -o 2 -mid 2 -tid 1 -ni 400 -qo 5 -nor -vl 1 -alc 0.5 -st 1 -no-vis
```
gives
```diff
-  Final strain energy: 0.009217 = metrics: 0.00858223 + limiting term: 0.000634768
+  Final strain energy: 0.009217 = metrics: 0.009217 + limiting term: 0
```

**Pull request merged**